### PR TITLE
Sim coil sensitivities

### PR DESCRIPTION
**Pull request by ckolbPTB**
This numerical simulation of a birdcage coil is taken from: https://github.com/ismrmrd/ismrmrd-python-tools/blob/17fe6fbf1bb645112e2ad0023eb4f42afb36421e/ismrmrdtools/simulation.py#L53

We can then use this simulation to generate test cases for the csm-methods of #64, #34 and #5

### CODE REVIEW

File: src/mrpro/phantoms/_CoilSensitivities.py [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+        ----------
+            relative_radius: Relative radius of birdcage
+
+        This function is strongly inspired by https://github.com/ismrmrd/ismrmrd-python-tools. The associated license
+        information can be found at the end of this file.
+        """
+
+        dim = [self.number_of_coils, self.image_dimensions.y, self.image_dimensions.x]
+        sensitivities = torch.zeros(dim, dtype=torch.complex64)
+        y_co, x_co = torch.meshgrid(
```
Comment (fzimmermann89) [inline src/mrpro/phantoms/_CoilSensitivities.py]: x_co and y_co are inthe wrong order IMHO:

```
dim = [1, 5, 6]  # c y x
y_co, x_co = torch.meshgrid(
    torch.linspace(-dim[2] // 2, dim[2] // 2 - 1, dim[2]), #x
    torch.linspace(-dim[1] // 2, dim[1] // 2 - 1, dim[1]), #y
    indexing='xy',
)
print(y_co.shape) # torch.Size([5, 6])
print(y_co[:,0])  # tensor([-3., -3., -3., -3., -3.])
print(y_co[0,:])  # tensor([-3., -2., -1.,  0.,  1.,  2.])
```
-> y_co currently changes along the last dim, which would be x in our naming convention

File: src/mrpro/phantoms/_CoilSensitivities.py [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+        image_dimensions
+            Number of voxels in the image
+        normalize_with_rss
+            If set to true, the calculated sensitivities are normalized by the root-sum-of-squares
+    """
+
+    def __init__(
+        self,
+        number_of_coils: int,
+        image_dimensions: SpatialDimension[int],
```
Comment (fzimmermann89) [inline src/mrpro/phantoms/_CoilSensitivities.py]: If we take a spatialdimension for the dimension, I would prefer that the return value is always of shape
(1,ncoils,nz.ny.nx), or broadcastable to this. It is surprising, that even though we can supply an z dimension, it will just be ignored and not match the expected shape.

So instead for a 2D birdcage, we would always return (1,ncoils,1, ny.nx).

File: src/mrpro/phantoms/_CoilSensitivities.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+
+        dim = [self.number_of_coils, self.image_dimensions.y, self.image_dimensions.x]
+        sensitivities = torch.zeros(dim, dtype=torch.complex64)
+        y_co, x_co = torch.meshgrid(
+            torch.linspace(-dim[2] // 2, dim[2] // 2 - 1, dim[2]),
+            torch.linspace(-dim[1] // 2, dim[1] // 2 - 1, dim[1]),
+            indexing='xy',
+        )
+
+        for c in range(0, dim[0]):
```
Comment (fzimmermann89) [inline src/mrpro/phantoms/_CoilSensitivities.py]: This could maybe also be vectorized over the coils without the loop?

**Comment by fzimmermann89** [comment]
I am not sure if this should be a class...
It just looks like a function? What is the advantage over having a birdcage function?

Why are image dimensions init parameters?

This might also be a question of preference, if something can be a pure function, I prefer to have it as a function.
So you can call coilsensitivities.birdcage(dim, width) instead of Coilsensititive(dim).birdcage(width).
In my head, objects are things with state, or "things", not collections of functions. this is what modules / .py files are for.

The only reason I could so to have it as a class, would be to have a CoilSensitivies(QData) class(as we plan to do)  and with a class method simulate_birdcage() that returns a simulated cls instance.


Remark for the future: We should at some point add random (within a given range) diameter, random phase (currently, the phase increases linearly with the coil number), and random orientation to be able to use this for NN training.

Also, there might be a bug with the meshgrid., see inline comment

**Comment by ckolbPTB** [comment]
> I am not sure if this should be a class...
> It just looks like a function? What is the advantage over having a birdcage function?

Good point. It can also be a function.

> The only reason I could so to have it as a class, would be to have a CoilSensitivies(QData) class(as we plan to do) and with a class method simulate_birdcage() that returns a simulated cls instance.

I think I prefer keeping that separate. We will have a CsmData(QData) class but I would prefer to use something like
```
csm_sim = coils.birdcage()
csm = CsmData.from_tensor_and_header(csm_sim, header)
```
rather than
`csm = CsmData.simulate_birdcage()`
because this would suggest that we also have 
`IData.simulate_ellipse()` and `KData.simulate_ellipse()`

File: src/mrpro/phantoms/_CoilSensitivities.py (hunk 2 above)
Comment (ckolbPTB) [reply src/mrpro/phantoms/_CoilSensitivities.py]: Agreed. Would it make more sense to have
1) image_dimensions: SpatialDimension[int] and return (1,ncoils,1, ny, nx)
or
2) have nx, ny as input (as we do for EllipsePhantom) and return (ncoils, ny, nx)?

**Comment by fzimmermann89** [comment]
>I think I prefer keeping that separate.

I concur 👍

Should we (once we have CsmData implemented) change it, such that the function returns CsmData instead of a tensor, i.e.
 ```
csm:CsmData = phantoms.coils.birdcage(header, **othersettings)
csm_sim:Tensor = csm.data #if we need the tensor
 ```
and similar for the other phantoms?

Or will all of our phantom / simulation data return "naked" tensors that we then have to wrap?
 (Sorry if  we already have discussed and decided this when merging the ElipsePhantom)
I think I would be fine with either, if we make it a "rule"

Also, we should be consistent with regard to the SpatialDimension/(ny,nx) in the phantoms module.
Once again, I would be fine with either, as I see the advantage of having the phantoms a bit more "naked" such that they can be easily used outside our packages as well as the advantage of defining x,y properly.

What about accepting a SpatialDimension or a Tuple with order z,y,x ? Is this "too flexible" for our taste?

**Comment by ckolbPTB** [comment]
> Or will all of our phantom / simulation data return "naked" tensors that we then have to wrap?
> (Sorry if we already have discussed and decided this when merging the ElipsePhantom)
> I think I would be fine with either, if we make it a "rule"

I would prefer to keep it as simple tensors. If we pass a header to the simulation functions then it is not immediately obvious to the user which parameters of the header are used and need to be changed in order to modify the simulation. 

> What about accepting a SpatialDimension or a Tuple with order z,y,x ? Is this "too flexible" for our taste?

I see the problem here that we introducted SpatialDimension in order to avoid having to "remember" in which order we provide the dimensions. Allowing now a tuple with order z,y,x seems like taking a step back. I would prefer to either use `birdcage_2d(ncoils, idim: SpatialDimension)` or `birdcage_2d(ncoils, ny, nx)`. 

I am happy with both - one small advantage of the second option is to me that it is clearer to the user what is used. In the first case should we throw an error if idim[2] > 1?

File: src/mrpro/phantoms/_CoilSensitivities.py (hunk 2 above)
Comment (fzimmermann89) [reply src/mrpro/phantoms/_CoilSensitivities.py]: 3. have  `ncoils=1,nz=1, ny=1, nx=1` as input and return (1, ncoils, 1,ny,nx) ?
this way, all coil sims take the same signature, regardless of 2d/3d?

4.  have nx,ny,nz keyword-only arguments
```
def birdcage(*, nz=1, ny, nx, **othersettings): #
    print(nz, ny, nx)
birdcage(nx=1, ny=2, nz=3)  #3 2 1
birdcage(nx=3, ny=2) # 1 2 3
birdcage(nz=2, nx=1) # invalid, as ny is missing
birdcage(1,2,3) # confusing, what is the order. but invalid as nz,ny,nx are keyword-only arguments
```

Disadvantage of 2.):
we would need to have some special logic in` CSMData.__init__(data:tensor, header:QHeader)`  to be able to to` CSMData(csm_sim,header=header)` . 
Or we would have to do ` CSMData(csm_sim[None,:,None],header=header)` (or similar).

So I'd prefer to output (1, ncoils, 1,ny,nx). 

Not sure between 1 and 3 and if we should have an nz input for a 2D simulation.
Advantage: Same signature for different simulations
Disadvantage: Might be confusing why there is a nz if it is not used, and why birdcage(10,2,3) returns (ncoils=10,nz=1,ny=3,nx=1)

If we do nx,ny,nz as separate inputs, we might consider making them keyword-only arguments.
Advantage: order is clear
Disadvantage: more to write whenever we call it


Really not sure... 

(Either way, I assume we would at some point switch the EllipsePhantom to the same logic)

**Comment by fzimmermann89** [comment]
> In the first case should we throw an error if idim[2] > 1?

I would always return (1, ncoils, nz=1,ny,nx) as this is broadcastable to the requested shape without throwing an error.

**Comment by ckolbPTB** [comment]
Should we also change `EllipsePhantom.image_space(ny, nx)` to `EllipsePhantom.image_space(SpatialDimension)`?

**Comment by fzimmermann89** [comment]
> Should we also change `EllipsePhantom.image_space(ny, nx)` to `EllipsePhantom.image_space(SpatialDimension)`?


Yes, good idea.
If you want to, you could include this as a second commit into this PR (i.e. squash the coil-related into one commit and the ellipse-related changes into one commit) or as a small second PR, whatever is easier for you.
(Or leave this as a todo if you don't want to do it yourself)

File: src/mrpro/phantoms/coils.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+            Relative radius of birdcage
+        normalize_with_rss
+            If set to true, the calculated sensitivities are normalized by the root-sum-of-squares
+
+    This function is strongly inspired by https://github.com/ismrmrd/ismrmrd-python-tools. The associated license
+    information can be found at the end of this file.
+    """
+
+    dim = [number_of_coils, image_dimensions.y, image_dimensions.x]
+    sensitivities = torch.zeros(dim, dtype=torch.complex64)
```
Comment (fzimmermann89) [inline src/mrpro/phantoms/coils.py]: not required, will be overwritten in l61

File: src/mrpro/phantoms/coils.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+    rr = torch.sqrt((x_co[None, ...] - coil_center_x) ** 2 + (y_co[None, ...] - coil_center_y) ** 2)
+    phi = torch.arctan2((x_co[None, ...] - coil_center_x), -(y_co[None, ...] - coil_center_y)) + coil_phase
+    sensitivities = (1 / rr) * np.exp(1j * phi)
+
+    if normalize_with_rss:
+        rss = torch.sqrt(torch.sum(torch.abs(sensitivities) ** 2, 0))
+        # Normalize only where rss is > 0
+        sensitivities[:, rss > 0] /= rss[None, rss > 0]
+
+    return rearrange(sensitivities, '(other coil z) y x->other coil z y x', other=1, z=1)
```
Comment (fzimmermann89) [inline src/mrpro/phantoms/coils.py]: could be
 `einops.repeat(sensitivities, 'coil y x->other coil z y x', other=1, z=1)`
  as I recently learned and prefer.
 
  I think I used rearrange like you did also in a couple of places and will at some point replace them with repeats, so  doesn't matter right now, but maybe for the future

**Review by fzimmermann89 — APPROVED** [review]
2 tiny changes that are not really important

**Comment by ckolbPTB** [comment]
@fzimmermann89 could you please have another quick check I did not mess up anything with the SpatialDimensions? Thanks!

**Comment by fzimmermann89** [comment]
lgtm
